pypeit/pypeit.py

```diff
@@ -417,102 +303,233 @@ def reduce_exposure(self, frame, bgframe=None, std_frame=None, reuse_masters=Fal
+                    std_trace = std_trace.flatten()
+                elif 'Echelle' in self.spectrograph.pypeline:
+                    std_trace = std_trace.T
+                else:
+                    msgs.error('Unrecognized pypeline')
+            else:
+                std_trace = None
+        else:
+            std_trace = None
+
+        return std_trace
+
+    def extract_one(self, frames, det, bg_frames=[], std_outfile=None):
+        """
+        Extract a single exposure/detector pair
+
+        sci_ID and det need to have been set internally prior to calling this method
+
+        Returns:
+            sciimg
+            sciivar
+            skymodel
+            objmodel
+            ivarmodel
+            outmask
+            sobjs
+            vel_corr
+
+        """
+        # Grab some meta-data needed for the reduction from the fitstbl
+        self.objtype, self.setup, self.obstime, self.basename, self.binning = self.get_sci_metadata(frames[0], det)
+        # Is this an IR reduction
+        self.ir_redux = True if len(bg_frames) > 0 else False
+        # Is this a standard star?
+        self.std_redux = 'standard' in self.objtype
+        # Get the standard trace if need be
+        std_trace = self.get_std_trace(self.std_redux, det, std_outfile)
+        # Instantiate ScienceImage for the files we will reduce
+        self.sciI = scienceimage.ScienceImage(self.caliBrate.tslits_dict, self.spectrograph,
+                                              self.fitstbl.frame_paths(frames),
+                                              bg_file_list=self.fitstbl.frame_paths(bg_frames),
+                                              ir_redux = self.ir_redux,
+                                              par=self.par['scienceimage'],
+                                              frame_par=self.par['scienceframe'],
+                                              objtype=self.objtype,
+                                              det=det,
+                                              binning=self.binning,
+                                              setup=self.setup)
+        # For QA on crash
+        msgs.sciexp = self.sciI
+
+        # Process images (includes inverse variance image, rn2 image, and CR mask)
+        sciimg, sciivar, rn2img, mask, crmask = \
+            self.sciI.proc(self.caliBrate.msbias, self.caliBrate.mspixflatnrm,
+                           self.caliBrate.msbpm, illum_flat=self.caliBrate.msillumflat,
+                           show=self.show)
+        # Object finding, first pass on frame without sky subtraction
+        self.maskslits = self.caliBrate.maskslits.copy()
+        # Do one iteration of object finding, and sky subtract to get initial sky model
+        self.sobjs_obj, self.nobj, skymask_init = \
+            self.find_objects(sciimg, std=self.std_redux, ir_redux=self.ir_redux, std_trace=std_trace, snr_trim=False, maskslits=self.maskslits)
```

Comment:
I find this odd to be in the main PypeIt class.

Would make more sense in a child or (better)
in ScienceImage.

Reply:
Which part do you mean the object finding? I think it is not as straightforward as you think to move this functionality all to the Scienceimage class, since there is overlap with PypeIt. I guess we need to think about where we want the main control flow of the reduction to be?

Reply:
The reason I didn't put the code in children of PypeIt for the control flow is that the code is identical for both, so it would just be replicating code. Originally there were separate methods, but then I realized they could be made identical with minimal inheritance of an object finding method.